bitcoin/BRSyncManager.c

```diff
@@ -0,0 +1,1549 @@
+
+static void
+BRPeerSyncManagerSubmit(BRPeerSyncManager manager,
+                        OwnershipGiven BRTransaction *transaction);
+
+static void
+BRPeerSyncManagerTickTock(BRPeerSyncManager manager);
+
+// PeerManager callbacks
+static void _BRPeerSyncManagerSyncStarted (void *info);
+static void _BRPeerSyncManagerSyncStopped (void *info, int reason);
+static void _BRPeerSyncManagerTxStatusUpdate (void *info);
+static void _BRPeerSyncManagerSaveBlocks (void *info, int replace, BRMerkleBlock **blocks, size_t count);
+static void _BRPeerSyncManagerSavePeers  (void *info, int replace, const BRPeer *peers, size_t count);
+static int  _BRPeerSyncManagerNetworkIsReachabele (void *info);
+static void _BRPeerSyncManagerThreadCleanup (void *info);
+static void _BRPeerSyncManagerTxPublished (void *info, int error);
+
+/// MARK: - Sync Manager Implementation
+
+extern BRSyncManager
+BRSyncManagerNewForMode(BRSyncMode mode,
+                        BRSyncManagerEventContext eventContext,
+                        BRSyncManagerEventCallback eventCallback,
+                        BRSyncManagerClientContext clientContext,
+                        BRSyncManagerClientCallbacks clientCallbacks,
+                        OwnershipKept const BRChainParams *params,
+                        OwnershipKept BRWallet *wallet,
+                        uint32_t earliestKeyTime,
+                        uint64_t blockHeight,
+                        OwnershipKept BRMerkleBlock *blocks[],
+                        size_t blocksCount,
+                        OwnershipKept const BRPeer peers[],
+                        size_t peersCount) {
+    switch (mode) {
+        case SYNC_MODE_BRD_ONLY:
+        return BRSyncManagerNewForClient (eventContext,
+                                          eventCallback,
+                                          clientContext,
+                                          clientCallbacks,
+                                          params,
+                                          wallet,
+                                          earliestKeyTime,
+                                          blockHeight,
+                                          blocks,
+                                          blocksCount,
+                                          peers,
+                                          peersCount);
+        case SYNC_MODE_P2P_ONLY:
+        return BRSyncManagerNewForP2P (eventContext,
+                                       eventCallback,
+                                       params,
+                                       wallet,
+                                       earliestKeyTime,
+                                       blockHeight,
+                                       blocks,
+                                       blocksCount,
+                                       peers,
+                                       peersCount);
+        default:
+        assert (0);
+        return NULL;
+    }
+}
```

Comment:
Can you sketch how this is going to work for the other modes?  [Other modes are arguably not in this first release; but at least SYNC_MODE_BRD_WITH_P2P_SEND is coming along right fast.]

Reply:
We would have a new SyncManager implementation. How it would work internally we'd determine at that time. Trying to plan for that now seems like a guess at best. We can (and likely will) refactor at that time.

Reply:
But that is what I'm asking - just for a flavor of how it might work.  It seems to me that the other modes will have a bit of `ForClient()` and a bit of `ForP2P`.  It is almost like the wallet manager always needs `ForClient()` and then for BTC and ETH has an optional `ForP2P`.